pinecone/core/grpc/index_grpc.py

```diff
@@ -155,42 +160,160 @@ def wrapped():
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.close()
+
+
+class vector(object):
+    def __init__(self, id: str, values: list, metadata: dict):
+        self.id = id
+        self.values = values
+        self.metadata = metadata
+
+    def __repr__(self):
+        return str({'id': self.id, 'values': self.values, 'metadata': self.metadata})
+
+
+class FetchResponse(object):
+    def __init__(self, vectors: dict, namespace: str):
+        self.vectors = vectors
+        self.namespace = namespace
+
+    def __repr__(self):
+        return str({'vectors': self.vectors, 'namespace': self.namespace})
+
+
+def parse_fetch_response(response: dict):
+    vd = {}
+    vectors = response.get('vectors')
+    if not vectors:
+        return None
+    for id, vec in vectors.items():
+        v_obj = vector(id=vec['id'], values=vec['values'], metadata=vec.get('metadata', None))
+        vd[id] = v_obj
+    namespace = response.get('namespace', None)
+    f = FetchResponse(vectors=vd, namespace=namespace)
+    return f
+
+
+class ScoredVector(object):
+    def __init__(self, id: str, score: float, values: list, metadata: dict):
+        self.id = id
+        self.score = score
+        self.values = values
+        self.metadata = metadata
 
-class Index(GRPCIndex):
+    def __repr__(self):
+        return str({'id': self.id, 'score': self.score, 'values': self.values, 'metadtata': self.metadata})
+
+
+class QueryResult(object):
+    def __init__(self, matches: list, namespace: str):
+        self.matches = matches
+        self.namespace = namespace
+
+    def __repr__(self):
+        return str({'matches': self.matches, 'namespace': self.namespace})
+
+
+class QueryResponse(object):
+    def __init__(self, results: list):
+        self.results = results
+
+    def __repr__(self):
+        return str({'results': self.results})
+
+
+def parse_query_response(response: dict):
+    res = []
+
+    for match in response['results']:
+        namespace = match.get('namespace', None)
+        m = []
+        for item in match['matches']:
+            sc = ScoredVector(id=item['id'], score=item['score'], values=item.get('values', []),
+                              metadata=item.get('metadata', {}))
+            m.append(sc)
+        res.append(QueryResult(matches=m, namespace=namespace))
+    return QueryResponse(results=res)
+
+
+class GRPCIndex(GRPCIndexBase):
 
     @property
     def stub_class(self):
         return VectorServiceStub
 
-    def upsert(self,
-               request: 'vector_service_pb2.UpsertRequest',
-               timeout: int = None,
-               metadata: Dict[str, str] = None):
-        return self._wrap_grpc_call(self.stub.Upsert, request, timeout=timeout, metadata=metadata)
+    @sentry
+    @validate_and_convert_errors
+    def upsert(self, vectors, **kwargs):
+        def _vector_transform(item):
+            if isinstance(item, Vector):
+                item.metadata
```

Comment:
typo line

Reply:
@rajat08 this PR was merged although this line is still here